src/pkg/extensions/extensions.go

```diff
@@ -0,0 +1,126 @@
+package extensions
+
+import (
+	"context"
+	"io"
+
+	"github.com/alcionai/clues"
+
+	"github.com/alcionai/corso/src/pkg/backup/details"
+	"github.com/alcionai/corso/src/pkg/logger"
+)
+
+// Extension client interface
+type CorsoItemExtension interface {
+	io.ReadCloser
+}
+
+type CorsoItemExtensionFactory func(
+	context.Context,
+	io.ReadCloser,
+	details.ItemInfo,
+	*details.ExtensionInfo,
+) (CorsoItemExtension, error)
+
+// Thin wrapper for runtime logging & metrics around extensions
+type loggerExtension struct {
+	info    details.ItemInfo
+	innerRc io.ReadCloser
+	ctx     context.Context
+	extInfo *details.ExtensionInfo
+}
+
+func NewLoggerExtension(
+	ctx context.Context,
+	rc io.ReadCloser,
+	info details.ItemInfo,
+	extInfo *details.ExtensionInfo,
+) (CorsoItemExtension, error) {
+	return &loggerExtension{
+		ctx:     ctx,
+		innerRc: rc,
+		info:    info,
+		extInfo: extInfo,
+	}, nil
+}
+
+func (l *loggerExtension) Read(p []byte) (int, error) {
+	n, err := l.innerRc.Read(p)
+	if err != nil && err != io.EOF {
+		logger.CtxErr(l.ctx, err).Error("inner read")
+		return n, err
+	}
+
+	if err == io.EOF {
+		logger.Ctx(l.ctx).Debug("corso extensions: EOF")
```

Comment:
Let me address a few questions around verbose logging &  `loggingExtension` ROI here.

I was thinking in terms of E2E observability. Since extension lifetimes are per item, it's going to generate some logging chaf. 

Specifically, we should be able to answer below questions using logs. 

1. If the extensions were supplied, were they enabled?
2. If the extensions were enabled, did they run?
3. Did extensions run for all items, or just some of them? If some items were skipped, was it corso at fault or was it the extension?
4. If a read/close failed, was it because of extension code or corso code?
5. If the extensions produced data, was it persisted by corso?

Having a `loggingExtension` wrapper can help us answer 2, 3, and to a certain extent 4.

Additionally, my north star for adding `loggingExtension` was to have an upper `loggingExtension` & lower `loggingExtension` with third party extensions sandwiched in between. It can help us isolate metrics around extension runtimes, enforce read only checks ( compare lower.crc with upper.crc) etc. No current need here though.

Reply:
> If the extensions were supplied, were they enabled?

IMO, every extension that's supplied should be enabled.  The caller can decide what gets enabled or not from within the extension implementation.

> If the extensions were enabled, did they run?

Same as above, that shouldn't be corso's concern (outside of basic smoke testing).  

> Did extensions run for all items, or just some of them? If some items were skipped, was it corso at fault or was it the extension?

io.Readers give us a lot of stability here that alleviates fault from corso.  One of the reasons for pushing toward a design centralized around wrapping the reader interface instead of middle-manning copies of buffers is that we, in corso, cannot be "at fault".  At least, not beyond any behavior that would also fundamentally break corso.

> If a read/close failed, was it because of extension code or corso code?

Given the nature of golang errors, we'd find out either way by receiving the error at the topmost reader (corso, in this case).  That error should contain a record of its causes from which we can extract reviewable details.

> If the extensions produced data, was it persisted by corso?

In controlled testing, that assertion is better verified by review of the details entries post-processing.  Yes, there's a lot of assertions we want to make about the behavior.  But those are all assertions we can make in a generic case, by focusing on the system handling of extensions, not the external user's application of them.

Overall, again, I think the instinct here is good.  But it's a little early in the project for us to need this scale of granularity.  Especially when it comes to details like `compare lower.crc with upper.crc`.

I honestly won't be mad if you keep it in.  I may grumble a bit when I go to filter out the messages.  But ultimately it will add more value than it detracts.  I'd only ask that we defer to either Debug or Error logging, and avoid Info.